pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -175,67 +166,127 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
+                "Caught exception while re-calculating the target assignment: " + e1, instancePartitionsMap,
+                targetAssignment);
+          }
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while updating IdealState for table: {}", tableNameWithType, e);
+          return new RebalanceResult(RebalanceResult.Status.FAILED, "Caught exception while updating IdealState: " + e,
               instancePartitionsMap, targetAssignment);
         }
       }
-      // If min available replicas is negative, treat it as max unavailable replicas
-      if (minAvailableReplicas < 0) {
-        minAvailableReplicas = Math.max(numReplicas + minAvailableReplicas, 0);
+    }
+
+    // Calculate the min available replicas for no-downtime rebalance
+    int minAvailableReplicas = rebalanceConfig.getInt(RebalanceConfigConstants.MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME,
+        RebalanceConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME);
+    int numCurrentReplicas = currentAssignment.values().iterator().next().size();
+    int numTargetReplicas = targetAssignment.values().iterator().next().size();
+    // Use the smaller one to determine the min available replicas
+    int numReplicas = Math.min(numCurrentReplicas, numTargetReplicas);
+    if (minAvailableReplicas > 0) {
+      if (minAvailableReplicas >= numReplicas) {
+        LOGGER.warn(
+            "Illegal config for min available replicas: {} for table: {}, must be less than number of replicas (current: {}, target: {})",
+            minAvailableReplicas, tableNameWithType, numCurrentReplicas, numTargetReplicas);
+        return new RebalanceResult(RebalanceResult.Status.FAILED, "Illegal min available replicas config",
+            instancePartitionsMap, targetAssignment);
       }
-      LOGGER.info("Rebalancing table: {} with min available replicas: {}", tableNameWithType, minAvailableReplicas);
     }
+    // If min available replicas is negative, treat it as max unavailable replicas
+    if (minAvailableReplicas < 0) {
+      minAvailableReplicas = Math.max(numReplicas + minAvailableReplicas, 0);
+    }
+    LOGGER.info("Rebalancing table: {} with min available replicas: {}", tableNameWithType, minAvailableReplicas);
 
     int expectedVersion = currentIdealState.getRecord().getVersion();
     while (true) {
       // Wait for ExternalView to converge before updating the next IdealState
+      IdealState idealState;
       try {
-        IdealState idealState = waitForExternalViewToConverge(tableNameWithType);
-        LOGGER.info("ExternalView converged for table: {}", tableNameWithType);
-        if (idealState.getRecord().getVersion() != expectedVersion) {
-          LOGGER.info(
-              "IdealState version changed while waiting for ExternalView to converge for table: {}, re-calculating the target assignment",
-              tableNameWithType);
+        idealState = waitForExternalViewToConverge(tableNameWithType);
```

Comment:
Should we be passing the expected version into `waitForExternalViewToConverge` so that if the idealstate changes due to some other factor, we don't wait for convergence. Instead, we should execute the logic below.

Reply:
We still need to wait for convergence when ideal state changes to guarantee the available replicas